CPS-427: Civicase alignment with CiviCRM 5.35

reneolivo opened the pull request:
## Overview
This PR fixes a number of bugs that are currently present on Civicase in order to properly run it in CiviCRM version 5.35.

**Note:** This PR also includes:
* https://github.com/compucorp/uk.co.compucorp.civiawards/pull/208
* https://github.com/compucorp/uk.co.compucorp.civicrm.prospect/pull/75

## Bug fixes

Note, it's important to understand that apart from core changes, AngularJS has also been upgraded to version 1.8 and this contributes to some of the issues we see.

### CiviCRM Dashboard

The CiviCRM Dashboard was breaking because it has been completely redone in AngularJS. Our current implementation that added an activity tab to the dashboard is no longer valid.

We decided to remove the Dashboard for now and implement it at a later ticket since this requires more planning.

#### Previous Dashboard
![pr](https://user-images.githubusercontent.com/1642119/109815206-c2370b00-7c05-11eb-8829-79178fa1c20a.gif)


#### Current Dashboard
![pr](https://user-images.githubusercontent.com/1642119/109815381-f4e10380-7c05-11eb-9448-427369e93ddf.gif)

### Technical details

All files related to the CRM Dashboard we implemented have been removed.

### Unknown path error
When navigating to any Civicase page we get a "unknown path" message and no actual case management content.

#### Before
<img width="1125" alt="Screen Shot 2021-02-28 at 9 09 47 PM" src="https://user-images.githubusercontent.com/1642119/109440627-4ff9d700-7a09-11eb-9b45-52e8d58acb1f.png">

#### After
<img width="1227" alt="Screen Shot 2021-02-28 at 9 10 55 PM" src="https://user-images.githubusercontent.com/1642119/109440678-7750a400-7a09-11eb-9f18-30d4dee29cca.png">

#### Technical details
This issue is because the CiviCRM Angular modules could not find the Civicase JS files.

There have been many recent changes to the asset manager:
https://github.com/civicrm/civicrm-core/pull/18247 (as an example, not a specific problem). One of these changes have made it so that full paths to a file are not longer found by core. 

ie:
```php
return [
  'js' => [
    'ang/civicase/*.js', // CiviCRM glob works ✅ 
    'ang/relative/path/to/file.js', // works ✅ 
    '/full/path/to/file.js', // doesn't work 🚫 
  ],
];
```

To fix this issue we updated the `GlobRecursive` class so it returns files relatively to the extension's path. We updated all relevant angular modules so they use the `getRelativeToExtension` method.

### Manage cases errors

When visiting the manage cases page, we get AngularJS infinite digest loop errors.

#### Before
<img width="1634" alt="Screen Shot 2021-02-28 at 9 39 27 PM" src="https://user-images.githubusercontent.com/1642119/109442056-74f04900-7a0d-11eb-9153-25ccfbcab500.png">

#### After
<img width="1227" alt="Screen Shot 2021-02-28 at 9 10 55 PM" src="https://user-images.githubusercontent.com/1642119/109440678-7750a400-7a09-11eb-9f18-30d4dee29cca.png">

#### Technical details
The problem happens because the AngularJS [Strict Contextual Escaping service](https://docs.angularjs.org/api/ng/service/$sce)'s methods now return a new reference each time they are called. We were using `getActivityFeedUrl` (which uses `$sce.trustAsResourceUrl` under the hood) directly on templates and since this would return a new reference the digest would not stop.

To stop this we removed the `$sce.trustAsResourceUrl` from `getActivityFeedUrl`. We have done a quick smock test and this has not broken any major area. The reason for using `$sce.trustAsResourceUrl` was also never documented.


### Broken Activities tab
The Activities tab is missing the archive list on the right sidebar.

#### Before
<img width="1485" alt="Screen Shot 2021-02-28 at 10 17 13 PM" src="https://user-images.githubusercontent.com/1642119/109444161-eed70100-7a12-11eb-87bb-e1efb888303c.png">

#### After
<img width="1490" alt="Screen Shot 2021-02-28 at 10 18 33 PM" src="https://user-images.githubusercontent.com/1642119/109444164-f4344b80-7a12-11eb-977e-967bc66cc753.png">

#### Technical details
The problem was that the `Activity.Getmonthswithactivities` endpoint was breaking because the `_civicrm_api3_get_BAO(__FUNCTION__)` call was not returning the `CRM_Activity_BAO_Activity` BAO class. This is probably related to the refactoring that was done to split  `civicrm_api3_activity_Getmonthswithactivities` into multiple functions to make them more readable, but now `_civicrm_api3_get_BAO(__FUNCTION__)` was not able to determine the BAO class we needed using the name of the function. We tried renaming the function but this did not work. 

Instead of relying on the function name to determine the BAO file we are 100% sure that we need the `CRM_Activity_BAO_Activity` BAO class we pass it directly instead of using `_civicrm_api3_get_BAO`. This makes the function more clear.

### Failing unit tests

Some of the unit tests failed after upgrading to the new Angular version.

* Civicase's repo test actions have been updated to use the new CiviCRM 5.35 version.
* Mocked `$broadcast` events needed to have a `callThrough` appended. Here is a good explanation for the fix: https://stackoverflow.com/questions/22721657/typeerror-cannot-read-property-defaultprevented-of-undefined
* After we removed the `$sce.trustAsResourceUrl` method from `getActivityFeedUrl` we should have removed the test with the expectation. As a note, there was a PR that added this test, but again, no explanation why `$sce.trustAsResourceUrl` was needed since the solution works without it just fine.
* Workflow tests were failing because we were trying to add a controller to the `/caseType/:id` route, but this route was never defined so all workflow tests were failing. We created a mock route and updated the karma configuration file so it include mocks and mock modules before unit tests run.
* One of the broadcast tests was always succeeding in our previous implementation because no `$broadcast` events were triggered at all. Now that `$broadcast` events are triggering it started failing so we defined the specific `$broadcast` event that we are not expecting.
* Some workflow tests were failing because of an unhandled promise. We patched this using `      $qProvider.errorOnUnhandledRejections(false);` on the spec file, but we need to confirm if we are properly handling rejections on the implementation.
* Some PHP unit tests were failing because the `CRM_Civicase_APIHelpers_CaseDetails` class was calling an internal private method statically from a static function, but the private method was not static. We changed the method so it's static.

deb1990 commented:
@reneolivo Great job. 
But please
1. Mention why we removed the Dashboard.
2. Technical Details for "Broken People's tab dates" is missing.

reneolivo commented:
@deb1990 thanks a lot. Regarding your comments:

1 - added why the Dashboard was removed.
2 - Removed the "Broken People's tab dates" section because that fix was related to files that were removed in this PR: https://github.com/compucorp/uk.co.compucorp.civicase/pull/714 since the solution doesn't need the files we updated here we just removed this commit.

lisandro-compucorp reviewed (approved):
👍
